standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -3531,15 +3612,18 @@ private LockResponse checkLock(Connection dbConn, long extLockId)
       Savepoint save = dbConn.setSavepoint();
       StringBuilder query = new StringBuilder("select hl_lock_ext_id, " +
         "hl_lock_int_id, hl_db, hl_table, hl_partition, hl_lock_state, " +
-        "hl_lock_type, hl_txnid from HIVE_LOCKS where hl_db in (");
+        "hl_lock_type, hl_txnid, hl_catalog from HIVE_LOCKS where hl_db in (");
 
       Set<String> strings = new HashSet<>(locksBeingChecked.size());
 
       //This the set of entities that the statement represented by extLockId wants to update
       List<LockInfo> writeSet = new ArrayList<>();
+      String catName = null;
 
       for (LockInfo info : locksBeingChecked) {
         strings.add(info.db);
+        if (catName == null) catName = info.catName;
+        else assert catName.equals(info.catName) : "Cross catalog locking operations are not allowed";
```

Comment:
This seems like the wrong place to enforce this, especially as an assert.  Shouldn't this be somewhere in the compiler that checks to ensure no query is using resources from more than 1 catalog?
Or perhaps in TxnManager for when we get to multi-stmt txns

Reply:
Agreed.